### PR TITLE
Accumulation of changes to RRTMGP

### PR DESCRIPTION
**Pull request by dustinswales**
## Description
This PR contains improvements and bug fixes to the RRTMGP enabled ccpp-physics.

- Bug fix (indexing error) for temperature being passed to GP. (Only exercised in rare cases in EMC's C768L127 configuration).
- New coupling of RRTMGP to the RRFS physics. Based on fv3_regional_quilt using FV3_GFS_v15_thompson_mynn_RRTMGP.
- Revised approach to the LW flux adjustment using the full vertical profile of the Jacobian of the upwelling LW. Previously only the surface flux was adjusted.
- Move allocation of GP interstitial DDTs from the LW scheme to GFS_typedefs.F90 interstitial_create/reset.

### Issue(s) addressed
N/A

## Testing
This was tested on Hera using the Intel compilers. There are no changes to existing baselines. A new baseline needs to be added for the new RT, fv3_regional_quilt_RRTMGP. 

Edit from @climbfuji:
- ALL RRTMGP tests change results
- Because of the bugfix for jet in ccpp-physics (previously, the compiler flag adjustments in `FV3/ccpp/physics/CMakeLists.txt` were ignored), most if not all results on jet change
- A number of bugfixes were required, some of which only showed up with the GNU compiler

Regression testing on tier-1 platforms:

- [x] hera.intel
- [x] hera.gnu
- [ ] orion.intel - skip because of ongoing filesystem (`/work`) problems
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [X] gaea.intel - log was lost due to bug in auto-rt, attaching the rt_auto log below instead.
- [x] jet.intel
- [x] wcoss_cray
- [x] wcoss_dell_p3

[gaea_intel_rt_auto_20210429173006.log](https://github.com/ufs-community/ufs-weather-model/files/6401106/gaea_intel_rt_auto_20210429173006.log)

## Dependencies
https://github.com/NOAA-EMC/fv3atm/pull/275
https://github.com/NCAR/ccpp-physics/pull/614

### CODE REVIEW

**Comment by RobertPincus** [comment]
Maybe just to emphasize that the title of the PR describes one change but the PR includes at least four changes.

File: tests/parm/v16_c96_rrtmgp.nml.IN [hunk 1]
```diff
@@ -199,7 +199,7 @@ deflate_level=1
   sw_file_gas    = 'rrtmgp-data-sw-g224-2018-12-04.nc'
   sw_file_clouds = 'rrtmgp-cloud-optics-coeffs-sw.nc'
   doG_cldoptics = .true.
-  rrtmgp_ngauss_ang = 3
+  rrtmgp_ngauss_ang = 1
```
Comment (RobertPincus) [inline tests/parm/v16_c96_rrtmgp.nml.IN]: @dustinswales Do we want to include this change in this PR? If so should we not use the same setting in every namelist, e.g. the new one you've provided for RRFS?

Comment (dustinswales) [reply tests/parm/v16_c96_rrtmgp.nml.IN]: @RobertPincus
Good catch. This did not belong.

**Review by Qingfu-Liu — APPROVED** [review]

**Review by Qingfu-Liu — APPROVED** [review]

File: tests/fv3_conf/gfs_v15_run.IN [hunk 2]
```diff
@@ -20,4 +20,5 @@ cp    @[INPUTDATA_ROOT]/FV3_input_data/*configure .
 
 if [ $DO_RRTMGP = .T. ]; then
 cp    @[INPUTDATA_ROOT]/FV3_input_data_RRTMGP/* .
+cp /scratch2/BMC/ome/Dustin.Swales/UFS/ufs-weather-model-2/FV3/ccpp/physics/physics/rte-rrtmgp/rrtmgp/data/* .
```
Comment (climbfuji) [inline tests/fv3_conf/gfs_v15_run.IN]: We need to create a new input data set and you need to update the input data set date tag on `rt.sh`. Let's use the same date for the new input data set and the new baseline, 20210428.

Then update this line.

File: tests/fv3_conf/gfs_v16_run.IN [hunk 3]
```diff
@@ -30,4 +30,5 @@ cp    @[INPUTDATA_ROOT]/FV3_input_data_INCCN_aeroclim/aer_data/LUTS/optics_SU.v1
 
 if [ $DO_RRTMGP = .T. ]; then
   cp @[INPUTDATA_ROOT]/FV3_input_data_RRTMGP/* .
+  cp /scratch2/BMC/ome/Dustin.Swales/UFS/ufs-weather-model-2/FV3/ccpp/physics/physics/rte-rrtmgp/rrtmgp/data/* .
```
Comment (climbfuji) [inline tests/fv3_conf/gfs_v16_run.IN]: To update

File: tests/fv3_conf/gsd_run.IN [hunk 4]
```diff
@@ -68,3 +68,7 @@ if [ $GWD_OPT = 3 ] || [ $GWD_OPT = 33 ] || [ $GWD_OPT = 2 ] || [ $GWD_OPT = 22
   cp @[INPUTDATA_ROOT]/FV3_input_data_gsd/drag_suite/* INPUT/
 fi
 
+if [ $DO_RRTMGP = .T. ]; then
```
Comment (climbfuji) [inline tests/fv3_conf/gsd_run.IN:71]: To update

File: tests/fv3_conf/regional_run.IN [hunk 5]
```diff
@@ -42,3 +42,8 @@ elif [[ $POSTAPP = 'hafs' ]]; then
     cp ${PATHRT}/parm/postxconfig-NT-hafs.txt postxconfig-NT_FH00.txt
     cp ${PATHRT}/parm/params_grib2_tbl_new params_grib2_tbl_new
 fi
+
+if [ $DO_RRTMGP = .T. ]; then
+  cp @[INPUTDATA_ROOT]/FV3_input_data_RRTMGP/* .
+  cp /scratch2/BMC/ome/Dustin.Swales/UFS/ufs-weather-model-2/FV3/ccpp/physics/physics/rte-rrtmgp/rrtmgp/data/* .
```
Comment (climbfuji) [inline tests/fv3_conf/regional_run.IN]: To update

File: tests/parm/v16_c96_rrtmgp.nml.IN [hunk 6]
```diff
@@ -41,14 +41,14 @@ deflate_level=1
 /
 
 &fv_core_nml
-  layout = @[INPES],@[JNPES]
+  layout = 3,8
```
Comment (climbfuji) [inline tests/parm/v16_c96_rrtmgp.nml.IN]: Why this change, isn't `3,8` the default value in `default_vars.sh` ?

File: tests/tests/fv3_gfs_v16_RRTMGP [hunk 7]
```diff
@@ -38,6 +44,12 @@ export LIST_FILES="atmos_4xdaily.tile1.nc \
                    dynf024.tile4.nc \
                    dynf024.tile5.nc \
                    dynf024.tile6.nc \
+                   dynf048.tile1.nc \
```
Comment (climbfuji) [inline tests/tests/fv3_gfs_v16_RRTMGP:47]: This won't work - how do you want to compare 48 output files if the runtime is only 24 hours?

**Review by climbfuji — CHANGES_REQUESTED** [review]

File: tests/tests/fv3_gfs_v16_RRTMGP (hunk 7 above)
Comment (dustinswales) [reply tests/tests/fv3_gfs_v16_RRTMGP:47]: Sorry, I added +export FHMAX=48 +export RESTART_INTERVAL=24 to make the GP v16 RTs just as for RRTMG.

Comment (climbfuji) [reply tests/tests/fv3_gfs_v16_RRTMGP:47]: Does the job still complete within 30 minutes?

File: tests/parm/v16_c96_rrtmgp.nml.IN (hunk 6 above)
Comment (dustinswales) [reply tests/parm/v16_c96_rrtmgp.nml.IN]: reverted.

File: tests/tests/fv3_gfs_v16_RRTMGP (hunk 7 above)
Comment (dustinswales) [reply tests/tests/fv3_gfs_v16_RRTMGP:47]: yes

File: tests/fv3_conf/gfs_v15_run.IN (hunk 2 above)
Comment (climbfuji) [reply tests/fv3_conf/gfs_v15_run.IN]: @dustinswales you can remove all occurrences of `cp /scratch2/BMC/ome/...` - Dusan and I copied the contents of that directory into `@[INPUTDATA_ROOT]/FV3_input_data_RRTMGP/` for the new input data set 20210428 on all tier-1 platforms.

Comment (dustinswales) [reply tests/fv3_conf/gfs_v15_run.IN]: Done.

File: tests/tests/fv3_gfs_v16_RRTMGP [hunk 8]
```diff
@@ -75,10 +87,14 @@ export LIST_FILES="atmos_4xdaily.tile1.nc \
 export_fv3
 
 DT_ATMOS="1200"
+export FHMAX=48
```
Comment (junwang-noaa) [inline tests/tests/fv3_gfs_v16_RRTMGP:90]: May I ask why the forecast time is increase from 24 to 48hr? Will 24hr work?

File: tests/tests/fv3_gfs_v16_RRTMGP [hunk 9]
```diff
@@ -75,10 +87,14 @@ export LIST_FILES="atmos_4xdaily.tile1.nc \
...
+export RESTART_INTERVAL=24
+
+export NSTF_NAME=2,1,0,0,0
```
Comment (junwang-noaa) [inline tests/tests/fv3_gfs_v16_RRTMGP:93]: Is this NSTF change required for the RRTMGP test? this is different from gfs_v16 configuration, which uses 2,0,0,0,0.

File: tests/tests/fv3_gfs_v16_RRTMGP (hunk 8 above)
Comment (dustinswales) [reply tests/tests/fv3_gfs_v16_RRTMGP:90]: I did this to be consistent with the fv3_gfs_v16 RT.

File: tests/tests/fv3_gfs_v16_RRTMGP (hunk 9 above)
Comment (dustinswales) [reply tests/tests/fv3_gfs_v16_RRTMGP:93]: I did this to be consistent with the fv3_gfs_v16 RT.

File: tests/tests/fv3_gfs_v16_RRTMGP (hunk 8 above)
Comment (junwang-noaa) [reply tests/tests/fv3_gfs_v16_RRTMGP:90]: fv3_gfs_v16 has this to prepare restart test. I am updating global regression tests, so the RRTMGP test will also be updated, will ask you to check when the PR is ready.

**Comment by BrianCurtis-NOAA** [comment]
Machine: jet
Compiler: intel
Job: BL
Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/612732341/20210428221509/ufs-weather-model
Please make changes and add the following label back:
jet-intel-BL

**Comment by BrianCurtis-NOAA** [comment]
Machine: gaea
Compiler: intel
Job: BL
Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/612732341/20210428221511/ufs-weather-model
Please manually delete: /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_2224
Repo location: /lustre/f2/pdata/ncep/emc.nemspara/autort/pr/612732341/20210428232529/ufs-weather-model
Please manually delete: /lustre/f2/scratch/emc.nemspara/FV3_RT/rt_3990
Test fv3_gfs_v16_RRTMGP_2thrd 047 failed in check_result failed 
Test fv3_gfs_v16_RRTMGP_2thrd 047 failed in run_test failed 
Please make changes and add the following label back:
gaea-intel-BL

**Comment by climbfuji** [comment]
Update on these PRs:
- several errors in the coderegression tests were addressed during the code review and while generating baselines yesterday
- two RRTMGP regression tests were commented out due to failures, among them the RRTMGP threading test
- the code does not compile on jet, and while we have created baselines on all other platforms we cannot proceed without knowing how the fix for jet impacts the code and the (RRTMGP) regression test results
- Note: because changes were made to detecting OpenMP for CCPP, @climbfuji compared the runtime of the existing control threading test (`fv3_2threads`) with the current trunk, and the runtime is identical (i.e. good)

The UFS code managers discussed, if the PR is not finalized and merged this afternoon MT, it will be postponed and we move on to the next PR.

@junwang-noaa @DusanJovic-NOAA @dustinswales @RobertPincus

**Comment by dustinswales** [comment]
@climbfuji 
The GP RT w/ threading worked for me. I had to make a change to rt.conf for the test to work. Working on the c192l127test now

**Comment by climbfuji** [comment]
> @climbfuji
> The GP RT w/ threading worked for me. I had to make a change to rt.conf for the test to work. Working on the c192l127test now

You need to revert this change. Adding `fv3` means that you are overwriting the single-thread results with the two-thread results. The 2thread test must pass against the single thread results.

More info: the `fv3` column is used to determine whether a test is run when baselines are created. Tests without `fv3` are only run when results are verified against existing baselines.

**Comment by dustinswales** [comment]
@

> > @climbfuji
> > The GP RT w/ threading worked for me. I had to make a change to rt.conf for the test to work. Working on the c192l127test now
> 
> You need to revert this change. Adding `fv3` means that you are overwriting the single-thread results with the two-thread results. The 2thread test must pass against the single thread results.
> 
> More info: the `fv3` column is used to determine whether a test is run when baselines are created. Tests without `fv3` are only run when results are verified against existing baselines.

Gotcha. Will revert.

**Comment by climbfuji** [comment]
Note: CI tests passed for commit a9a3887.

**Comment by climbfuji** [comment]
Regression tests passed on all tier-1 platforms, except Orion: filesystem issues `/work/` not responding, cannot run tests. I will start the merge with ccpp-physics.

**Review by climbfuji — APPROVED** [review]
I fixed the submodule pointer for fv3atm, ready to merge.

**Review by junwang-noaa — APPROVED** [review]

**Comment by dustinswales** [comment]
> I fixed the submodule pointer for fv3atm, ready to merge.

Thanks!
